### PR TITLE
Guard rotating refresh-token writes with compare-and-swap

### PR DESCRIPTION
**Pull request by reyortiz3**
## Summary

- Multiple replicas of an application embedding this auth server, sharing the same Redis-backed `UpstreamTokenStorage`, can redeem and persist the same single-use, rotating upstream refresh token concurrently (some OAuth providers issue short-lived access tokens together with single-use refresh tokens that rotate on every redemption). The existing `singleflight` dedup in `upstreamTokenRefresher` only coordinates refreshes within one process, and `StoreUpstreamTokens` is an unconditional overwrite — so a losing replica's write can silently clobber a winning replica's already-rotated refresh token, permanently breaking the refresh chain and forcing the user back through re-authentication.
- Adds `UpstreamTokenStorage.CompareAndSwapUpstreamTokens(ctx, sessionID, providerName, expectedRefreshToken, tokens)`, which only writes if the refresh token currently stored still equals `expectedRefreshToken`, returning the new `ErrConcurrentRefresh` sentinel otherwise. Implemented for both backends: a new CAS Lua script (`casUpstreamTokensScript`) for `RedisStorage`, and a mutex-guarded compare-then-write for `MemoryStorage`.
- `upstreamTokenRefresher.refreshAndStore` now writes through the CAS method (expected value = the refresh token it just redeemed with the provider) instead of the old unconditional `StoreUpstreamTokens`. On `ErrConcurrentRefresh` it re-reads the row: if another replica's write left it unexpired, that replica's tokens are returned to the caller; otherwise it surfaces a clear error rather than retrying an already-dead redemption. The re-read also distinguishes a genuine lost race (unexpired row) from a row that's simply gone (`ErrNotFound` — deleted by logout or evicted by TTL), logging the latter at `Warn` rather than `Error` so it isn't mistaken for a race in production logs.
- `singleflight` remains a same-process optimization (avoids redundant upstream calls / Redis round-trips for concurrent requests inside one process). The CAS write is what makes the *stored row* deterministic across processes — it is **not** a guarantee that concurrent redemption is safe at the upstream provider: both replicas still call the provider before either write lands, so a provider enforcing strict single-use rotation can still see two redemptions of the same refresh token regardless of which write wins here. CAS is fully sufficient only where the provider tolerates a grace/leeway window in which more than one redeemed child stays valid; closing the gap for a stricter provider needs a lock around the whole read-redeem-write sequence, which is a real follow-up, not implemented here (see below).

## Type of change

- [x] Bug fix

## Test plan

- [x] Unit tests (`task test`)
- [x] Linting (`task lint-fix`)

Added:
- `TestMemoryStorage_CompareAndSwapUpstreamTokens` / `TestRedisStorage_CompareAndSwapUpstreamTokens` — matching/stale/absent-row CAS cases, a security-positive case (CAS refuses to resurrect a row deleted between read and write, unlike `StoreUpstreamTokens`), plus a 10-goroutine concurrent-write race against a real (miniredis-backed) Lua script proving exactly one write wins and losers never clobber it.
- `TestUpstreamTokenRefresher_ConcurrentRefreshConflict` — the refresher's conflict-resolution path: returns the winning replica's tokens when the re-read is unexpired, and a wrapped `ErrConcurrentRefresh` error when no usable winner exists (both the row-absent and still-expired re-read shapes are covered).
- `TestUpstreamTokenRefresher_ConcurrentRefreshConflict_LogLevel` — proves the lost-race vs. row-absent distinction actually logs at different levels.

Updated all existing `refresher_test.go` mock expectations from `StoreUpstreamTokens` to `CompareAndSwapUpstreamTokens`. Mocks regenerated via `task gen`.

## API Compatibility

- [x] This PR does not break the `v1beta1` API, OR the `api-break-allowed` label is applied and the migration guidance is described above.

## Does this introduce a user-facing change?

No — this is an internal storage-layer/refresher change. It hardens an existing reliability gap; no public API changes beyond the additive `UpstreamTokenStorage` interface method.

## Special notes for reviewers

- `UpstreamTokenStorage.CompareAndSwapUpstreamTokens` is an additive interface method — any external implementation of this interface will need the new method to satisfy the interface (only `RedisStorage` and `MemoryStorage` implement it in this repo).
- No distributed lock (e.g. `SET NX` mutex) was introduced. CAS makes the stored row deterministic and prevents a stale write from clobbering a newer one, but for a provider that enforces strict single-use rotation with no grace window, a lock around the whole read-redeem-write sequence is a correctness requirement, not just an optimization — CAS alone doesn't stop both replicas from calling the provider before either writes. This matters concretely for Read.ai (the motivating provider): its access tokens are 10 minutes, so this race gets exercised often, and its refresh-token grace period is still an open item on its public roadmap (not documented as guaranteed today). Treating the follow-up lock as a real correctness item, not a nice-to-have.
- Not included in this PR: a hermetic OAuth test double modeling single-use rotating refresh tokens with a grace period, and a two-replica staging/e2e test against a real provider — left as potential follow-up work.

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/authserver/storage/redis.go [hunk 1]
```diff
@@ -1297,6 +1297,87 @@ end
 return 1
 `)
 
+// casUpstreamTokensScript is the compare-and-swap sibling of
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis.go:1311]: **Redis CAS script duplicates the store script's write/index body — consensus 10/10 (concurrency-correctness, go-style-interface-design)**

`casUpstreamTokensScript`'s write-and-index body (SET, session-index SADD/PEXPIRE/PERSIST invariants, user reverse-index SADD/SREM) is a verbatim copy of `storeUpstreamTokensScript`'s equivalent body, kept in sync only by a comment. A future fix to one script's TTL/index logic that isn't mirrored to the other will silently reintroduce a bug this codebase has already gone to real effort to get right once.

Consider unifying into one parameterized script (an ARGV sentinel meaning "no CAS check" for the unconditional path), or add a test asserting both scripts' shared body stays byte-identical so drift fails CI instead of relying on the comment.

File: pkg/authserver/refresher.go [hunk 2]
```diff
@@ -236,22 +262,73 @@ func (r *upstreamTokenRefresher) refreshAndStore(
...
+// overwrote it before this call's CAS ran), so the caller must surface an
+// error — retrying would only redeem the same already-consumed refresh token
+// again.
+func (r *upstreamTokenRefresher) resolveConcurrentRefreshConflict(
+	ctx context.Context, sessionID, providerID string,
+) (*storage.UpstreamTokens, error) {
+	readCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), refreshTimeout)
+	defer cancel()
+
+	winner, err := r.storage.GetUpstreamTokens(readCtx, sessionID, providerID)
```
Comment (tgrunnagle) [inline pkg/authserver/refresher.go:314]: **`resolveConcurrentRefreshConflict` discards the re-read error — consensus 9/10 (concurrency-correctness)**

`winner, err := r.storage.GetUpstreamTokens(...)` — on the unrecoverable path a few lines below, neither the `slog.Error` call nor the returned error includes `err`. A legitimately-expired/missing row and a transient re-read failure (e.g. a Redis timeout during recovery) are logged and returned identically, making an already-rare cross-process race much harder to diagnose from production logs.

Include `err` in the log call and wrap it into the returned error.

File: pkg/authserver/refresher.go [hunk 3]
```diff
@@ -194,7 +206,21 @@ func (r *upstreamTokenRefresher) refreshAndStore(
 		updated.IDToken = expired.IDToken
```
Comment (tgrunnagle) [inline pkg/authserver/refresher.go:221]: **Stale comment references the old unconditional write path — consensus 9/10 (go-style-interface-design)**

A few lines above this hunk (around the original line 189), a comment still reads "...StoreUpstreamTokens replaces the whole row, so without this the persisted IDToken would be overwritten..." — but this PR changes that exact call site to write through `CompareAndSwapUpstreamTokens` instead. The identical phrasing was already fixed in a test-file comment in this same PR, confirming this is an oversight rather than intentional.

Update the comment to reference `CompareAndSwapUpstreamTokens` (or phrase it generically as "the write call").

File: pkg/authserver/storage/redis_test.go [hunk 4]
```diff
@@ -2046,6 +2046,120 @@ func TestRedisStorage_UpstreamTokens(t *testing.T) {
...
+				go func(i int) {
+					defer wg.Done()
+					errs[i] = s.CompareAndSwapUpstreamTokens(ctx, "session", "provider-a", "shared-old-refresh", &UpstreamTokens{
+						AccessToken:  fmt.Sprintf("access-%d", i),
+						RefreshToken: fmt.Sprintf("refresh-%d", i),
+						ExpiresAt:    time.Now().Add(time.Hour),
+					})
+				}(i)
+			}
+			wg.Wait()
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis_test.go]: **Concurrent CAS test blocks on `wg.Wait()` with no timeout — consensus 9/10 (test-coverage)**

This repo's own testing rule ("Concurrent Tests: Always Add Timeouts to Blocking Barriers") requires a timeout/fail-fast path on blocking barriers like `WaitGroup.Wait()`. This new 10-goroutine race test calls it bare, unlike `TestRedisStorage_ConsumeAssertionJWT_Concurrent` elsewhere in this same file, which follows the rule. A regression in the Lua script's atomicity that caused a hang would stall this test until the global `go test` timeout instead of failing fast with a clear message.

Wrap this in the same timeout-guarded pattern used in `TestRedisStorage_ConsumeAssertionJWT_Concurrent` (or `refresher_test.go`'s `waitGroupTimeout` helper).

File: pkg/authserver/storage/memory_test.go [hunk 5]
```diff
@@ -1171,6 +1171,73 @@ func TestMemoryStorage_UpstreamTokens(t *testing.T) {
...
+		withStorage(t, func(ctx context.Context, s *MemoryStorage) {
+			err := s.CompareAndSwapUpstreamTokens(ctx, "session", "provider-a", "", &UpstreamTokens{
+				AccessToken: "first-access", RefreshToken: "first-refresh",
+			})
+			require.NoError(t, err)
+
+			retrieved, err := s.GetUpstreamTokens(ctx, "session", "provider-a")
+			require.NoError(t, err)
+			assert.Equal(t, "first-access", retrieved.AccessToken)
+		})
```
Comment (tgrunnagle) [inline pkg/authserver/storage/memory_test.go:1252]: **Missing CAS boundary test: empty expected value against a populated row — consensus 8/10 (test-coverage)**

Both backends test "empty expected value matches an absent row," but neither tests the inverse — `expectedRefreshToken=""` against a row that already has a non-empty refresh token. That's exactly the boundary that proves `""` can't be used to silently overwrite an already-populated row.

Add a subtest here (and the equivalent in `redis_test.go`) asserting `ErrConcurrentRefresh` and that the row is left untouched.

**Review by tgrunnagle — COMMENTED** [review]
## Multi-Agent Consensus Review
**Agents consulted**: concurrency-correctness, security, go-style-interface-design, test-coverage, general-code-quality

### Consensus Summary
| # | Finding | Consensus | Severity | Action |
|---|---------|-----------|----------|--------|
| 1 | Redis CAS script duplicates the store script's write/index body, sync-guarded only by a comment | 10/10 | MEDIUM | Fix |
| 2 | `resolveConcurrentRefreshConflict` discards the re-read error on the unrecoverable path | 9/10 | MEDIUM | Fix |
| 3 | Stale comment references `StoreUpstreamTokens` on a path that now writes through CAS | 9/10 | MEDIUM | Fix |
| 4 | New concurrent CAS test blocks on `wg.Wait()` with no timeout guard | 9/10 | MEDIUM | Fix |
| 5 | Architecture doc contradicts the new cross-process CAS guarantee | 8/10 | MEDIUM | Fix |
| 6 | No test for `expectedRefreshToken=""` against an already-populated row | 8/10 | MEDIUM | Fix |
| 7 | Untested "re-read succeeds but still expired" conflict-resolution branch | 8/10 | MEDIUM | Fix |
| 8 | Redis CAS script's session-index/TTL maintenance path has no dedicated test | 8/10 | MEDIUM | Fix |

### Overall
This closes a real correctness gap: multiple replicas sharing Redis-backed `UpstreamTokenStorage` could clobber each other's rotated single-use refresh tokens, since `StoreUpstreamTokens` was an unconditional overwrite. `CompareAndSwapUpstreamTokens` is the right fix — a CAS primitive rather than a heavier distributed lock — and it's implemented atomically in both backends, with the Redis path's atomicity proven by a genuine 10-goroutine race test against real miniredis rather than just a code read-through.

The findings here are all non-blocking. The most durable risk is the Redis CAS Lua script duplicating the entire write/index body of the pre-existing store script, guarded only by a "keep these in sync" comment — a future fix to one script's TTL/index logic that isn't mirrored to the other would silently reintroduce a bug this codebase already went to some effort to fix once. The conflict-resolution fallback path (`resolveConcurrentRefreshConflict`) also discards its re-read error, which will make an already-rare production race harder to diagnose, and its "still expired" branch (as opposed to "row not found") is untested. A few other test-coverage gaps round out the list: the CAS boundary case of an empty expected value against a populated row, and the new Redis script's index/TTL maintenance path for the CAS variant specifically.

None of this blocks merge — it's hardening on top of an already-sound fix.

### Documentation
`docs/arch/11-auth-server-storage.md`'s "Refresh Coordination Scope" section (lines 205-208) states: "This is deliberately narrower than \#4122: it does not provide distributed coordination across replicas, row-addressed mutation, compare-and-swap, or any other cross-process consistency guarantee." This PR adds exactly that guarantee and the doc should be updated to reflect it — the `refresher.go` package doc's "two layers" framing (in-process singleflight vs. cross-process CAS) is a good model to adapt from.

---
*Generated with Claude Code*

File: pkg/authserver/refresher_test.go [hunk 6]
```diff
@@ -891,3 +892,69 @@ func TestUpstreamTokenRefresher_RowIdentity(t *testing.T) {
...
+		provider := upstreammocks.NewMockOAuth2Provider(ctrl)
+
+		store.EXPECT().ResolveUpstreamTokenRowID(gomock.Any(), "session", "github").
+			Return(storage.UpstreamTokenRowID("row"), nil)
+		gomock.InOrder(
+			store.EXPECT().GetUpstreamTokens(gomock.Any(), "session", "github").Return(expired, storage.ErrExpired),
+			// The post-conflict re-read finds nothing usable: absent, or still
+			// expired. Either way this call cannot recover a usable token.
+			store.EXPECT().GetUpstreamTokens(gomock.Any(), "session", "github").
+				Return(nil, fmt.Errorf("%w: row evicted", storage.ErrNotFound)),
```
Comment (tgrunnagle) [inline pkg/authserver/refresher_test.go:950]: **Untested "still expired" conflict-resolution branch — consensus 8/10 (test-coverage)**

`resolveConcurrentRefreshConflict` treats a re-read error (tested here via `ErrNotFound`) and a re-read that succeeds but returns a still-expired row identically — but only the error case is exercised. The "still expired" case has a materially different `GetUpstreamTokens` return shape (`(someExpiredRow, nil)` vs. `(nil, err)`) and isn't covered by any test.

Add a third subtest for the `(expiredRow, nil)` re-read outcome.

File: pkg/authserver/storage/redis_test.go [hunk 7]
```diff
@@ -2046,6 +2046,120 @@ func TestRedisStorage_UpstreamTokens(t *testing.T) {
...
+			}))
+
+			err := s.CompareAndSwapUpstreamTokens(ctx, "session", "provider-a", "old-refresh", &UpstreamTokens{
+				AccessToken: "new-access", RefreshToken: "new-refresh", ExpiresAt: time.Now().Add(time.Hour),
+			})
+			require.NoError(t, err)
+
+			retrieved, err := s.GetUpstreamTokens(ctx, "session", "provider-a")
+			require.NoError(t, err)
+			assert.Equal(t, "new-access", retrieved.AccessToken)
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis_test.go:2070]: **Redis CAS script's index/TTL path is untested for the CAS variant — consensus 8/10 (test-coverage)**

`TestRedisStorage_UpstreamTokens` has dedicated assertions for the session-index/TTL maintenance behavior on the `StoreUpstreamTokens` path, but none of the new CAS subtests check `idxKey` membership, TTL, or the user-upstream set after a successful CAS write — exactly the code this PR's Lua script duplication (see the `redis.go` comment) puts at risk of silent drift.

Add one assertion here mirroring an existing `StoreUpstreamTokens` index check, as cheap insurance.

File: pkg/authserver/storage/redis.go (hunk 1 above)
Comment (reyortiz3) [reply pkg/authserver/storage/redis.go:1311]: Fixed. Extracted the write/index body into a shared `upstreamRowWriteAndIndexScriptBody` Go constant, concatenated into both `storeUpstreamTokensScript` and `casUpstreamTokensScript` — removes the duplication entirely rather than just guarding it with a comment.

File: pkg/authserver/refresher.go (hunk 2 above)
Comment (reyortiz3) [reply pkg/authserver/refresher.go:314]: Fixed. `err` is now included in both the log call and wrapped into the returned error.

File: pkg/authserver/refresher.go (hunk 3 above)
Comment (reyortiz3) [reply pkg/authserver/refresher.go:221]: Fixed. Reworded to "the write below" rather than naming a specific method.

File: pkg/authserver/storage/redis_test.go (hunk 4 above)
Comment (reyortiz3) [reply pkg/authserver/storage/redis_test.go]: Fixed. Wrapped in the same timeout-guarded done-channel pattern used by TestRedisStorage_ConsumeAssertionJWT_Concurrent in this file.

File: pkg/authserver/storage/memory_test.go (hunk 5 above)
Comment (reyortiz3) [reply pkg/authserver/storage/memory_test.go:1252]: Fixed. Added "empty expected value against an already-populated row returns ErrConcurrentRefresh" in both memory_test.go and redis_test.go.

File: pkg/authserver/refresher_test.go (hunk 6 above)
Comment (reyortiz3) [reply pkg/authserver/refresher_test.go:950]: Fixed. Added a third subtest, "winner still expired on re-read", covering the (expiredRow, nil) re-read shape distinct from the (nil, err) case already tested.

File: pkg/authserver/storage/redis_test.go (hunk 7 above)
Comment (reyortiz3) [reply pkg/authserver/storage/redis_test.go:2070]: Fixed. Added session-index-set membership/TTL and user-reverse-index assertions to the "matching expected value writes and returns nil" CAS subtest.

**Comment by reyortiz3** [comment]
Addressed all 8 findings from the multi-agent consensus review:

1. **Redis CAS script duplication** — unified `storeUpstreamTokensScript` and `casUpstreamTokensScript`'s write/index body into one shared `upstreamRowWriteAndIndexScriptBody` Go constant, concatenated into both scripts. Eliminates the duplication rather than just guarding it with a comment.
2. **Discarded re-read error** — `resolveConcurrentRefreshConflict` now includes `err` in both the log call and the returned (wrapped) error.
3. **Stale `StoreUpstreamTokens` comment** — reworded to reference the write call generically.
4. **Concurrent CAS test missing timeout guard** — wrapped in the same timeout-guarded done-channel pattern as `TestRedisStorage_ConsumeAssertionJWT_Concurrent` in this same file.
5. **Architecture doc contradicted the new guarantee** — updated `docs/arch/11-auth-server-storage.md`'s "Refresh Coordination Scope" section to describe the CAS layer alongside the existing in-process singleflight description.
6. **Missing empty-expected-value-against-populated-row test** — added to both `memory_test.go` and `redis_test.go`.
7. **Untested "still expired on re-read" branch** — added a third subtest distinguishing it from the "not found" case.
8. **Untested Redis CAS index/TTL path** — added session-index-set membership/TTL and user-reverse-index assertions to the successful-write CAS subtest.

All existing and new tests pass (`task test`), lint is clean (`task lint-fix`).

File: docs/arch/11-auth-server-storage.md [hunk 8]
```diff
@@ -202,10 +202,21 @@ distinct rows would cross credentials between sessions. This prevents
...
+`UpstreamTokenStorage.CompareAndSwapUpstreamTokens` conditions a refresh write
+on the refresh token currently stored still matching the value the caller
+redeemed with, failing the write (`ErrConcurrentRefresh`) instead of
+overwriting when another replica already rotated the row first. `refreshAndStore`
+writes through this method rather than an unconditional `StoreUpstreamTokens`.
+Redis implements the comparison and the write as one atomic Lua script;
+`MemoryStorage` implements it under its existing mutex. `singleflight` remains
+the process-local optimization described above — it avoids a redundant
+upstream call and Redis round-trip for concurrent requests inside one
+process — while the CAS write is what makes concurrent redemption safe across
```
Comment (amirejaz) [inline docs/arch/11-auth-server-storage.md]: I don't think this claim holds for every provider, and since it's landing in the architecture docs it's worth getting precise.

CAS orders writes to storage, so first writer wins. But the IdP's view of which token is live is set by the last successful redemption, and the second redemption still happens here because the CAS runs after `provider.RefreshTokens`. Those two orderings can disagree.

Concretely, A and B both read RT0. A redeems and gets RT1, B redeems and gets RT2. A's CAS wins and stores RT1, B's fails and RT2 is discarded. If the provider only honours the newest child, the live token at the IdP is RT2 and we have just persisted the dead one. And for a provider doing rotation-based replay detection per RFC 9700 section 4.14.2, B's redemption of an already-used RT0 is a replay: the AS cannot tell which party replayed, so it revokes the active refresh token and the grant is gone regardless of what we stored.

Where CAS is fully correct is the grace-period/leeway case, where both children stay valid.

So I would soften this to something like "the CAS write makes the stored row deterministic and prevents a stale write from clobbering a newer one", and drop "makes concurrent redemption safe across processes". Same for the note in the PR description that a lock is "an optimization, not a correctness requirement" - for a strict-rotation provider it is one. I'm fine with the lock staying a follow-up, I just don't want the docs to record the gap as closed.

File: pkg/authserver/storage/types.go [hunk 9]
```diff
@@ -65,6 +65,15 @@ var (
...
 	ErrUserNotProvisioned = errors.New("storage: user not provisioned")
+
+	// ErrConcurrentRefresh is returned by
+	// UpstreamTokenStorage.CompareAndSwapUpstreamTokens when the currently
+	// stored refresh token no longer equals the caller's expected value,
+	// meaning another writer (a concurrent refresh in this process or another
+	// replica sharing the same storage) already moved the row past it. The
+	// caller's own redemption is stale and must not be written; see
+	// CompareAndSwapUpstreamTokens for the coordination contract.
+	ErrConcurrentRefresh = errors.New("storage: upstream token row changed concurrently")
```
Comment (amirejaz) [inline pkg/authserver/storage/types.go:80]: This sentinel covers two fairly different situations: the row genuinely changed under us, and the row is no longer there. I checked both backends, and an absent row with a non-empty expected value returns `ErrConcurrentRefresh`:

```
MEMORY absent row, expected=non-empty -> storage: upstream token row changed concurrently
REDIS  absent row, expected=non-empty -> storage: upstream token row changed concurrently
```

Two things follow. The diagnostic one: a row evicted by TTL, or deleted by a logout between the read and the write, now logs "lost concurrent upstream token refresh race" at Error level, which sends whoever is on call chasing a concurrency bug that didn't happen. The lost-ack case does the same - attempt 1 lands, the client times out, attempt 2 mismatches against its own write, and the conflict path re-reads and recovers correctly but still logs a race.

The second is more interesting and I think worth calling out in the PR description. `StoreUpstreamTokens` would have resurrected a row that a logout deleted between the read and the write. CAS refuses to:

```
REDIS post-logout CAS -> storage: upstream token row changed concurrently
      row after       -> storage: item not found
```

That is a genuine security-relevant improvement, and it deserves a test, otherwise nothing stops someone from "fixing" it back later. A separate sentinel for the vanished-row case would cover both points.

File: pkg/authserver/storage/memory_test.go [hunk 10]
```diff
@@ -1171,6 +1171,93 @@ func TestMemoryStorage_UpstreamTokens(t *testing.T) {
...
+	})
+
+	t.Run("empty expected value against an already-populated row returns ErrConcurrentRefresh", func(t *testing.T) {
+		withStorage(t, func(ctx context.Context, s *MemoryStorage) {
+			require.NoError(t, s.StoreUpstreamTokens(ctx, "session", "provider-a", &UpstreamTokens{
+				AccessToken: "existing-access", RefreshToken: "existing-refresh",
+			}))
+
+			// "" must never be usable to silently overwrite a row that already
+			// carries a refresh token - it only matches a genuinely absent row.
```
Comment (amirejaz) [inline pkg/authserver/storage/memory_test.go]: This comment isn't quite right: `""` matches more than a genuinely absent row. It also matches a present row whose `refresh_token` is empty, and a null-marker row written from `nil` tokens. Both return nil rather than `ErrConcurrentRefresh`:

```
REDIS expected="" over null-marker row           -> <nil>
REDIS expected="" over present row with empty RT -> <nil>
```

The assertion itself is fine and the case is worth pinning. It's the "only matches a genuinely absent row" phrasing I'd change, because with the `attacker-*` naming here it reads as a guarantee that `""` can never overwrite a row that exists. The interface doc on `CompareAndSwapUpstreamTokens` is closer ("no row currently exists, or the existing row carries no refresh token"), though it misses the null-marker case too.

File: pkg/authserver/refresher.go [hunk 11]
```diff
@@ -236,22 +262,80 @@ func (r *upstreamTokenRefresher) refreshAndStore(
...
+// If the re-read is still expired, missing, or otherwise unusable, the loss is
+// unrecoverable for this attempt: the refresh token this call redeemed is now
+// dead both at the IdP (which enforces single-use) and in storage (the winner
+// overwrote it before this call's CAS ran), so the caller must surface an
+// error — retrying would only redeem the same already-consumed refresh token
+// again.
+func (r *upstreamTokenRefresher) resolveConcurrentRefreshConflict(
+	ctx context.Context, sessionID, providerID string,
+) (*storage.UpstreamTokens, error) {
+	readCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), refreshTimeout)
```
Comment (amirejaz) [inline pkg/authserver/refresher.go]: `ctx` here is already the detached `refreshCtx` from `RefreshAndStore`, which carries its own 30s `refreshTimeout`. Wrapping it in a fresh `WithoutCancel` plus another `refreshTimeout` gives this re-read a full 30s beyond the outer deadline, so the worst case for a caller blocked in singleflight goes from 30s to 60s, and neither bound is tied to the inbound request.

A single re-read doesn't need the same budget as an entire refresh. `upstreamDeleteTimeout` is 5s for much the same reason a couple of functions down; something in that range would fit better here.

**Review by amirejaz — COMMENTED** [review]
Read through this focusing on the concurrency contract and the OAuth side. The CAS mechanism itself looks well built, the shared Lua body from the last round is the right call, and `pkg/authserver/...` plus `pkg/auth/upstreamtoken/...` pass with `-race` on d88d643. Storage writes are now deterministic and a stale write can no longer clobber a newer one, which is a real fix.

My one substantive concern is the claim about what CAS guarantees, left inline on the arch doc: it orders storage writes, not redemptions at the IdP, and the second redemption still happens. That doesn't block the mechanism, but I'd like the wording in the docs and the PR description adjusted so the remaining gap is recorded rather than described as closed. The other three are smaller: an error sentinel that conflates "changed" with "gone", one inaccurate comment on a new test, and a timeout that doubles the worst-case wait.

File: docs/arch/11-auth-server-storage.md (hunk 8 above)
Comment (reyortiz3) [reply docs/arch/11-auth-server-storage.md]: Fixed. Softened the claim in this doc, the refresher.go type-level comment, the expectedRefreshToken comment, and the CompareAndSwapUpstreamTokens/ErrConcurrentRefresh interface docs to describe what CAS actually guarantees (a deterministic stored row) versus what it doesn't (safety against a strict single-use-rotation provider seeing two redemptions before either write lands). Called out that closing that gap needs a lock around the whole read-redeem-write sequence, tracked as a follow-up, not this PR.

File: pkg/authserver/storage/types.go (hunk 9 above)
Comment (reyortiz3) [reply pkg/authserver/storage/types.go:80]: Fixed, without a new sentinel: resolveConcurrentRefreshConflict now branches on errors.Is(err, storage.ErrNotFound) on the re-read and logs at Warn ("expected on logout or TTL eviction, not necessarily a lost race") instead of Error for that case, keeping Error only for the genuinely ambiguous case. Added a log-level test. Also added the security-positive test you flagged: CAS refuses to resurrect a row deleted between read and write (both backends) - and documented that behavior in the interface docs. Kept a single sentinel rather than adding a second one since no caller today needs to distinguish the two mechanically; the re-read (which RefreshAndStore already performs) is sufficient, and it's what the doc comment on ErrConcurrentRefresh now spells out.

File: pkg/authserver/storage/memory_test.go (hunk 10 above)
Comment (reyortiz3) [reply pkg/authserver/storage/memory_test.go]: Fixed. Reworded to note "" also matches a present row with its own empty RefreshToken or a nil-tokens row, and that this test only pins the has-a-real-refresh-token case. Also tightened the CompareAndSwapUpstreamTokens interface doc to spell out all three empty-match cases including the nil-tokens/null-marker one.

File: pkg/authserver/refresher.go (hunk 11 above)
Comment (reyortiz3) [reply pkg/authserver/refresher.go]: Fixed. Added a dedicated upstreamConflictReadTimeout (5s, same rationale as upstreamDeleteTimeout) and switched resolveConcurrentRefreshConflict to use it instead of the outer refreshTimeout.

**Comment by reyortiz3** [comment]
Addressed the second round of review feedback (4 findings from @amirejaz):

1. **Overclaimed "safe across processes"** — softened throughout (this doc, `refresher.go`'s type-level and inline comments, the `CompareAndSwapUpstreamTokens`/`ErrConcurrentRefresh` interface docs). CAS makes the *stored row* deterministic; it doesn't by itself make concurrent redemption safe against a strict single-use-rotation provider, since both replicas still call the provider before either write lands. Closing that gap needs a lock around the whole read-redeem-write sequence — noted as a real follow-up, not claimed as done.
2. **`ErrConcurrentRefresh` conflates "raced" vs "row genuinely gone"** — `resolveConcurrentRefreshConflict` now logs at `Warn` (not `Error`) when the re-read comes back `ErrNotFound`, since that's expected on logout/TTL eviction, not a lost race. Added a log-level test. Also added the security-positive test: CAS refuses to resurrect a row deleted between read and write (both backends), which the unconditional `StoreUpstreamTokens` would not have prevented.
3. **Imprecise "only matches a genuinely absent row" comment** — corrected; `""` also matches a present row with its own empty `RefreshToken` or a nil-tokens row, and the interface doc now spells out all three cases.
4. **Doubled worst-case timeout** — added a dedicated `upstreamConflictReadTimeout` (5s) instead of reusing the outer `refreshTimeout` (30s) for the post-conflict re-read.

All tests pass (`task test`), lint clean (`task lint-fix`).

File: pkg/authserver/storage/redis.go [hunk 12]
```diff
@@ -1295,7 +1306,49 @@ if newUserID ~= "" then
...
 
 return 1
-`)
+`
+
+// casUpstreamTokensScript is the compare-and-swap sibling of
+// storeUpstreamTokensScript: it additionally gates the write on the existing
+// row's refresh_token matching ARGV[5] before doing anything else, returning
+// 0 (no write performed) on a mismatch instead of 1. This is what makes
+// CompareAndSwapUpstreamTokens safe for redeeming a single-use, rotating
```
Comment (amirejaz) [inline pkg/authserver/storage/redis.go]: Small leftover from the last round: this still reads "is what makes `CompareAndSwapUpstreamTokens` safe for redeeming a single-use, rotating upstream refresh token across multiple replicas". The clause after the colon describes the real guarantee correctly, so it isn't misleading on its own, but "safe for redeeming" is the exact phrasing that got softened in the type-level comment, the interface doc and the arch doc. Worth matching them here so all four read the same way.

**Review by amirejaz — DISMISSED** [review]
Approving.

The concurrency contract holds up. CAS makes the stored row deterministic, a losing write can no longer clobber a rotated token, and the Redis comparison and write are genuinely one atomic operation now that both scripts share `upstreamRowWriteAndIndexScriptBody`. I re-ran `pkg/authserver/...` and `pkg/auth/upstreamtoken/...` with `-race` on 3f9ab97 and they're green.

Both things I raised last round are handled well. The `ErrNotFound` branch on the re-read is a better answer than the second sentinel I suggested: the double-`%w` means a caller can still tell the two situations apart (I checked - `errors.Is` reports both `ErrConcurrentRefresh` and `ErrNotFound`), so keeping one sentinel costs nothing. Adding the no-resurrection test in both backends was the right call too; that behavior was worth pinning before someone "fixes" it back.

Two doc-only leftovers, neither blocking.

The PR description still says a lock "would only save a redundant upstream redemption on the losing replica, which is an optimization, not a correctness requirement". Every code and doc site now says the opposite, so it's worth bringing the description in line before merge, since that's what stays visible in the merge commit later.

That's a bit sharper than it looks, given the arch doc now cites Read.ai as the provider where CAS is sufficient. Their access tokens are 10 minutes, so refreshes are frequent and this race gets exercised often, and their grace window isn't actually published - "refresh token grace period" is still an open item on their public roadmap. Worth linking the source in the doc, and I'd treat the follow-up lock as more than a nice-to-have.

The other leftover is inline on `redis.go`.

File: pkg/authserver/storage/redis.go (hunk 12 above)
Comment (reyortiz3) [reply pkg/authserver/storage/redis.go]: Fixed. Reworded to match the softened phrasing used everywhere else (the type-level comment, the interface doc, and the arch doc): CAS makes the stored row deterministic; it's not by itself a guarantee the redemption is safe at the provider.

**Comment by reyortiz3** [comment]
Fixed both doc-only leftovers:

1. Reworded the remaining "safe for redeeming" phrasing in the `casUpstreamTokensScript` comment (`redis.go`) to match the softened language used everywhere else.
2. Updated the PR description: the "optimization, not a correctness requirement" line now says the opposite (matching the code/docs), and added the Read.ai-specific context you raised — its 10-minute access tokens mean this race is exercised often, and its refresh-token grace period is still an open roadmap item rather than a documented guarantee. Treating the follow-up lock as a real correctness item now, not a nice-to-have.

**Review by amirejaz — APPROVED** [review]
